shared/actions/favorite.js

```diff
@@ -116,145 +163,102 @@ const folderToState = (folders: Array<FolderWithMeta>, username: string = ''): F
+  json.ignored && json.ignored.forEach(injectMeta('ignored'))
+  json.new && json.new.forEach(injectMeta('new'))
+
+  // figure out who can solve the rekey
+  const folders: Array<FolderWithMeta> = _jsonToFolders(json, myKID)
+
+  // Ensure private/public folders exist for us
+  if (username && loggedIn) {
+    [true, false].forEach(isPrivate => {
+      const idx = folders.findIndex(f => f.name === username && f.private === isPrivate)
+      let toAdd = {
+        meta: null,
+        name: username,
+        private: isPrivate,
+        notificationsOn: false,
+        created: false,
+        waitingForParticipantUnlock: [],
+        youCanUnlock: [],
+      }
 
-        if (folder.problem_set.can_self_help) {
-          const mySolutions = solutions[myKID] || []
-          folder.youCanUnlock = mySolutions.map(kid => {
-            const device = json.devices[kid]
-            return {...device, deviceID: kid}
-          })
-        } else {
-          folder.waitingForParticipantUnlock = Object.keys(solutions).map(userID => {
-            const devices = solutions[userID].map(kid => json.devices[kid].name)
-            const numDevices = devices.length
-            let last
-
-            if (numDevices > 1) {
-              last = devices.pop()
-            }
-
-            return {
-              name: json.users[userID],
-              devices: `Tell them to turn on${numDevices > 1 ? ':' : ' '} ${devices.join(', ')}${last ? ` or ${last}` : ''}.`,
-            }
-          })
-        }
+      if (idx !== -1) {
+        toAdd = folders[idx]
+        folders.splice(idx, 1)
       }
+
+      folders.unshift(toAdd)
     })
-  })
-  return _.flatten(folderSets)
+  }
+
+  return folders
 }
 
-export function favoriteList (): (dispatch: Dispatch, getState: () => Object) => void {
-  return (dispatch, getState) => {
-    // Ignore any messages we receive while reloading
-    if (getState().dev && getState().dev.reloading) {
-      return
-    }
+function * getFavoritesListSaga (): SagaGenerator<any, any> {
+  const bail = yield select(({dev: {reloading = false} = {}}) => reloading)
+  if (bail) {
+    return
+  }
 
-    apiserverGetRpc({
-      param: {
-        endpoint: 'kbfs/favorite/list',
-        args: [{key: 'problems', value: '1'}],
-      },
-      callback: (error, result) => {
-        if (error) {
-          console.warn('Err in getFavorites', error)
-          return
-        }
+  const results = yield call(_getFavoritesRPC)
+  // $ForceType
+  const {username, loggedIn} = yield select(
+    ({config: {username = null, loggedIn = false} = {}}) => ({username, loggedIn}))
 
-        let json
-        try {
-          json = JSON.parse(result.body)
-        } catch (err) {
-          console.warn('Invalid json from getFavorites: ', err)
-          return
-        }
+  // $ForceType
+  const state: FolderState = yield call(_folderToState, results && results.body, username, loggedIn)
```

Comment:
if this is a pure function, I think it makes sense to not use `call` because you lose typing, and you don't gain anything by having this be called outside this generator